arch/xtensa/core/mpu.c

```diff
@@ -0,0 +1,703 @@
+	}
+
+	/* Sort the entries in ascending order of starting address */
+	sort_entries(entries);
+
+	/*
+	 * Need to figure out where the start and end entries are as sorting
+	 * may change their positions.
+	 */
+	entry_slot_s = (struct xtensa_mpu_entry *)
+		       check_addr_in_mpu_entries(entries, start_addr, first_enabled_idx,
+						 &exact_s, &idx_s);
+	entry_slot_e = (struct xtensa_mpu_entry *)
+		       check_addr_in_mpu_entries(entries, end_addr, first_enabled_idx,
+						 &exact_e, &idx_e);
+
+	/* Both must be exact match. */
+	__ASSERT_NO_MSG(exact_s);
+	__ASSERT_NO_MSG(exact_e);
+
+	if (end_addr == 0xFFFFFFFFU) {
+		/*
+		 * If end_addr = 0xFFFFFFFFU, entry_slot_e and idx_e both
+		 * point to the last slot. Because the incoming region goes
+		 * to the end of memory, we simply cheat by including
+		 * the last entry by incrementing idx_e so the loop to
+		 * update entries will change the attribute of last entry
+		 * in map.
+		 */
+		idx_e++;
+	}
+
+	/*
+	 * Any existing entries between the "newly" popluated start and
+	 * end entries must bear the same attributes. So modify them
+	 * here.
+	 */
+	for (int idx = idx_s + 1; idx < idx_e; idx++) {
+		xtensa_mpu_entry_attributes_set(&entries[idx], access_rights, memory_type);
+	}
+
+	if (first_idx != NULL) {
+		*first_idx = first_enabled_idx;
+	}
+
+	ret = 0;
+
+out:
+	return ret;
+}
+
+/**
+ * Write the MPU map to hardware.
+ *
+ * @param map Pointer to foreground MPU map.
+ */
+static void xtensa_mpu_map_write(struct xtensa_mpu_map *map)
+{
+	int entry;
+
+	/*
+	 * Clear MPU entries first, then write MPU entries in reverse order.
+	 *
+	 * Remember that the boundary of each memory region is marked by
+	 * two consecutive entries, and that the addresses of all entries
+	 * must not be in descending order (i.e. equal or increasing).
+	 * To ensure this, we clear out the entries first then write them
+	 * in reverse order. This avoids any intermediate invalid
+	 * configuration with regard to ordering.
+	 */
+	for (entry = 0; entry < XTENSA_MPU_NUM_ENTRIES; entry++) {
+		__asm__ volatile("wptlb %0, %1\n\t" : : "a"(entry), "a"(0));
+	}
+
+	for (entry = XTENSA_MPU_NUM_ENTRIES - 1; entry >= 0; entry--) {
+		__asm__ volatile("wptlb %0, %1\n\t"
+				 : : "a"(map->entries[entry].at), "a"(map->entries[entry].as));
+	}
+}
+
+/**
+ * Perform necessary steps to enable MPU.
+ */
+void xtensa_mpu_init(void)
+{
+	unsigned int entry;
+	uint8_t first_enabled_idx;
+
+	/* Disable all foreground segments before we start configuration. */
+	xtensa_mpu_mpuenb_write(0);
+
+	/*
+	 * Clear the foreground MPU map so we can populate it later with valid entries.
+	 * Note that we still need to make sure the map is valid, and cannot be totally
+	 * zeroed.
+	 */
+	for (entry = 0; entry < XTENSA_MPU_NUM_ENTRIES; entry++) {
+		/* Make sure MBZ fields are cleared. */
+		xtensa_mpu_map_fg_kernel.entries[entry].as.raw = 0U;
+		xtensa_mpu_map_fg_kernel.entries[entry].at.raw = 0U;
```

Comment:
Style: this bit was confusing when I saw it the first time.  The "raw" field is in fact the full word and not the MBZ fields; I thought there was an uninitizlied data issue when there wasn't.  For clarity, suggest doing the work on a fully-initialized local variable instead and writing the whole thing into the table, e.g.:

```
struct xtensa_mpu_entry ent = {0};

ent.at.p.segment = entry;
...
...
xtensa_mpu_map_fg_kernel.entries[entry] = ent;
```

Reply:
With regard to `mpu_map_region_add()`, some of the complication is due to trying to figure out if there is an entry with the exact address as incoming region. If an entry exists, we update the entry. There is no guarantee on the ordering via `qsort()` for entries that are considered equal. So after `qsort()`, we have not idea which same address entries would be removed via `consolidate_entries()`.

As for the MBZ one, I have updated that to use your sample.